### PR TITLE
[Performance] Shared-memory Indices Permutation in DDP and _ScalarDataBatcher

### PR DESCRIPTION
**Pull request by yaox12**
## Description
This is to tackle issue #3357. We share the indices tensor among all the worker processes. At the beginning of each epoch, each worker process shuffles a continuous range of the shared indices tensor and then gets its own indices by stride slicing. This is not truly random but should be pretty good.

This should use O(N) memory and O(N/p) time for shuffling. However, I fail to capture the memory usage precisely due to noises. 

cc @BarclayII @nv-dlasalle 

## Performance

Run the script
```bash
python3 train_sampling_unsupervised.py --num-layers 1 --fan-out 8 --batch-size 500000 --gpu 0,1,2,3,4,5,6,7 --num-epochs 1 --dataset ogbn-products --data-cpu`
```
I got the time of shuffling indices tensor (average of processes) and enumerating the dataloader for the first time as below (in seconds):

|                    | shuffle | enum dataloader |
|--------------|---------|-----------------|
| dgl nightly   |    6.56 |            9.72 |
| randperm and assign  |    0.95 |            2.39 |
| shuffle a continuous range | 1.26 |        2.90 |

- `randperm and assign`: Implementation in this PR.
- `shuffle a continuous range`: It's something like the following code, we find it a bit slower.
```python
np.random.shuffle(self._shared_indices[self.num_samples*self.rank:self.num_samples*(self.rank + 1)].numpy())
```

## Known Issue

Currently, we use `dist.broadcast_object_list()`, which is introduced in PyTorch 1.7, to share the filename of `_shared_indices` among worker processes. We may need a workaround for lower PyTorch versions.

### CODE REVIEW

File: python/dgl/dataloading/pytorch/dataloader.py [hunk 1]
```diff
@@ -101,6 +101,32 @@ def __init__(self, dataset, shuffle=False, batch_size=1,
...
+                indices = indices.share_memory_()
+                _shared_filename = list(indices.storage()._share_filename_())
+            else:
+                _shared_filename = [None, None, None]
+            # Note: dist.broadcast_object_list() is a new feature introduced in PyTorch 1.7.0.
+            #       We may need a fallback for lower PyTorch versions.
+            dist.broadcast_object_list(_shared_filename, src=0)
+            # reconstruct the shared tensor
+            if self.rank != 0:
+                indices = th.LongTensor(th.LongStorage._new_shared_filename(*_shared_filename))
```
Comment (VoVAllen) [inline python/dgl/dataloading/pytorch/dataloader.py:125]: Interesting usage that I don't know before. Can we also use this to share the graph across multiple process? @BarclayII

**Review by VoVAllen — COMMENTED** [review]
The solution is quite neat. Thanks for the contribution. I'm not so familiar with this part and will let @BarclayII to review this.

File: python/dgl/dataloading/pytorch/dataloader.py (hunk 1 above)
Comment (BarclayII) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: Since they are prefixed with `_` I suspect that they are undocumented hence unstable.  Could we use a queue instead since it's the preferred way of sharing tensors across processes?

Comment (BarclayII) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: An alternative might be using `NDArray::EmptyShared` which creates a shared memory tensor directly.  The shared memory name can be generated in rank 0 and broadcasted to other ranks.

Comment (yaox12) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: >  use a queue

We have to create a queue and pass it to the worker processes when spawning/forking them, hard to do it inside the dataloader.

> using `NDArray::EmptyShared`

In fact I have tried it. `empty_shared_mem` uses a name to identify the shared memory. But I realize that there may be multiple dataloaders for training/test/validation, so we cannot use a hard code here.

@BarclayII

Comment (BarclayII) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: > `empty_shared_mem` uses a name to identify the shared memory. But I realize that there may be multiple dataloaders for training/test/validation, so we cannot use a hard code here.

That's because you use the same name `_shared_indices_for_ddp`?  I think you could generate a name in rank 0 and distribute it with `dist.broadcast` like you are doing here.

Comment (VoVAllen) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: I think it's fine to use this internal API since the semantic is clear. We can add some check or fallback solution if needed.
Our EmptyShared doesn't support windows as what I remembered. That will break the user experience on windows.

Comment (VoVAllen) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: Also I believe for error handling and error messages, Pytorch does a better job than us

Comment (BarclayII) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: You could also partition the training indices without shuffling to every rank, and then shuffle each partition individually.

Comment (BarclayII) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: @VoVAllen I can take care of supporting shared memory on Windows since they have a very similar interface.

Comment (yaox12) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: > You could also partition the training indices without shuffling to every rank, and then shuffle each partition individually.

Do you mean using a static split, i.e., the partition doesn't change between epochs? For this case, I think letting user partition the indices and initialize the dataloader with `use_ddp=False` will make things much easier.

If you mean something like:
```python
self._indices = _shared_indices[self.rank:self.total_size:self.num_replicas]
# then in iter_ddp():
np.random.shuffle(self._indices.numpy())
```
This would be even slower, which takes 2x times for shuffling compared to the above `randperm and assign`.

Comment (BarclayII) [reply python/dgl/dataloading/pytorch/dataloader.py:125]: > Do you mean using a static split, i.e., the partition doesn't change between epochs? For this case, I think letting user partition the indices and initialize the dataloader with `use_ddp=False` will make things much easier.

Right, say that we have N nodes and R workers, then each rank `r` is only responsible for shuffling the node ID `[N // R * r, N // R * (r+1)]`.  You could still use `randperm` and assign.  This should be OK because even if the nodes are ordered by label, since the gradients are synchronized across workers and would not be too far off.

Comment (yaox12) [reply python/dgl/dataloading/pytorch/dataloader.py]: @BarclayII I have pushed a version that uses a static indices split for each process. It's less random but being the fastest way.

**Comment by yaox12** [comment]
This is covered by the new data loader. So close it.
